File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/model/SubjectRepository.java
package com.ficture7.aasexplorer.model;

import com.ficture7.aasexplorer.Loader;
import com.ficture7.aasexplorer.Saver;
import com.ficture7.aasexplorer.model.examination.Examination;
import com.ficture7.aasexplorer.util.ObjectUtil;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import static com.ficture7.aasexplorer.util.ObjectUtil.checkNotNull;

/**
 * Represents a {@link Repository repository} of {@link Subject subjects}.
 *
 * @author FICTURE7
 */
public class SubjectRepository implements Repository<Integer, Subject> {

    //TODO: Create some form factory to create Subject & Resource instances.

    private boolean isLoaded;

    private final Examination examination;
    private final Class<? extends Examination> examinationClass;

    private final Loader loader;
    private final Saver saver;

    // Maps subject ID/codes to subject instances.
    private final Map<Integer, Subject> subjects;

    /**
     * Constructs a new instance of the {@link SubjectRepository} class with the
     * specified {@link Examination}, {@link Loader} and {@link Saver}.
     *
     * @param examination {@link Examination} instance.
     * @param loader      {@link Loader} instance.
     * @param saver       {@link Saver} instance.
     * @throws IllegalArgumentException {@code examination} is null.
     * @throws IllegalArgumentException {@code loader} is null.
     * @throws IllegalArgumentException {@code saver} is null.
     */
    public SubjectRepository(Examination examination, Loader loader, Saver saver) {
        this.examination = checkNotNull(examination, "examination");
        this.loader = checkNotNull(loader, "loader");
        this.saver = checkNotNull(saver, "saver");

        subjects = new HashMap<>();
        examinationClass = examination.getClass();
    }

    /**
     * Returns the {@link Subject} with the specified subject ID/code.
     *
     * @param subjectId Subject ID/code.
     * @return {@link Subject} with the specified subject ID/code.
     */
    @Override
    public Subject get(Integer subjectId) {
        return subjects.get(subjectId);
    }

    /**
     * Puts the specified {@link Subject} instance to the {@link SubjectRepository}.
     *
     * @param subject {@link Subject} instance.
     * @throws IllegalArgumentException {@code subject} is null.
     */
    public void put(Subject subject) {
        checkNotNull(subject, "subject");

        subjects.put(subject.id(), subject);
    }

    /**
     * Returns a boolean value indicating weather the {@link SubjectRepository} is loaded.
     *
     * @return A boolean value indicating weather the {@link SubjectRepository} is loaded.
     */
    @Override
    public boolean isLoaded() {
        return isLoaded;
    }

    /**
     * Loads the {@link SubjectRepository}.
     *
     * @throws Exception When an exception is thrown.
     */
    @Override
    public void load() throws Exception {
        // Unloads the repository to prevent duplication.
        unload();

        // Get the subject sources from the loader which can come from a
        // a client or a store.
        Iterable<SubjectSource> sources = loader.loadSubjects(examinationClass);

        for (SubjectSource source : sources) {
            Subject subject = get(source.id());

            if (subject != null) {
                subject.sources().add(source);
            } else {
                subject = new Subject(examination, loader, saver, source.id(), source.name());
                subject.sources().add(source);

                put(subject);
            }
        }

        isLoaded = true;
    }

    /**
     * Unloads the {@link SubjectRepository}.
     */
    @Override
    public void unload() {
        subjects.clear();
        isLoaded = false;
    }

    /**
     * Saves the {@link SubjectRepository}.
     *
     * @throws Exception When an exception is thrown.
     */
    @Override
    public void save() throws Exception {
        // Merge all the subject sources into a single linear list.
        List<SubjectSource> sources = new ArrayList<>();
        for (Subject subject : this) {
            for (SubjectSource source : subject.sources()) {
                sources.add(source);
            }
        }

        // Pass the subjects to save to the saver instance.
        saver.saveSubjects(examinationClass, sources);
    }

    /**
     * Returns an iterator through the {@link Subject} in the {@link SubjectRepository}.
     *
     * @return An iterator through the {@link Subject} in the {@link SubjectRepository}.
     */
    @Override
    public Iterator<Subject> iterator() {
        return subjects.values().iterator();
    }
}

File: aasexplorer-core/src/main/java/com/ficture7/aasexplorer/model/SubjectRepository.kt
package com.ficture7.aasexplorer.model

import com.ficture7.aasexplorer.Loader
import com.ficture7.aasexplorer.Saver
import com.ficture7.aasexplorer.model.examination.Examination

import java.util.ArrayList
import java.util.HashMap

/**
 * Represents a [repository][Repository] of [subjects][Subject].
 *
 * @author FICTURE7
 */
class SubjectRepository : Repository<Int, Subject> {

    //TODO: Create some form factory to create Subject & Resource instances.

    private var _isLoaded: Boolean = false

    private val factory: Subject.Factory
    private val examination: Examination
    private val examinationClass: Class<Examination>

    private val loader: Loader
    private val saver: Saver

    // Maps subject ID/codes to subject instances.
    private val subjects: MutableMap<Int, Subject>

    /**
     * Constructs a new instance of the [SubjectRepository] class with the
     * specified [Examination], [Loader] and [Saver].
     *
     * @param examination [Examination] instance.
     * @param loader      [Loader] instance.
     * @param saver       [Saver] instance.
     * @throws NullPointerException `examination` is null.
     * @throws NullPointerException `getLoader` is null.
     * @throws NullPointerException `getSaver` is null.
     */
    internal constructor(examination: Examination, loader: Loader, saver: Saver) {
        this.examination = examination
        this.loader = loader
        this.saver = saver

        subjects = HashMap()
        factory = Subject.Factory(examination, loader, saver)
        examinationClass = examination.javaClass
    }

    /**
     * Returns a boolean value indicating weather the [SubjectRepository] is loaded.
     *
     * @return A boolean value indicating weather the [SubjectRepository] is loaded.
     */
    override val isLoaded: Boolean
        get() = _isLoaded

    /**
     * Returns the [Subject] with the specified subject ID/code.
     *
     * @param subjectId Subject ID/code.
     * @return [Subject] with the specified subject ID/code.
     */
    override fun get(subjectId: Int): Subject? {
        return subjects[subjectId]
    }

    /**
     * Puts the specified [Subject] instance to the [SubjectRepository].
     *
     * @param subject [Subject] instance.
     */
    override fun put(subject: Subject) {
        subjects.put(subject.id, subject)
    }

    /**
     * Loads the [SubjectRepository].
     *
     * @throws Exception When an exception is thrown.
     */
    @Throws(Exception::class)
    override fun load() {
        // Unloads the repository to prevent duplication.
        unload()

        // Get the subject getSources from the getLoader which can come from a
        // a getClient or a getStore.
        val sources = loader.loadSubjects<Examination>(examinationClass)

        for (source in sources) {
            var subject: Subject? = get(source.id)

            if (subject != null) {
                subject.sources.add(source)
            } else {
                subject = factory.create(source.id, source.name)
                subject.sources.add(source)

                put(subject)
            }
        }

        _isLoaded = true
    }

    /**
     * Loads the [SubjectRepository] asynchronously.
     *
     * @param callback Callback initializer.
     */
    override fun loadAsync(callback: Repository.LoadCallback.() -> Unit) {
        val loadCallback = Repository.LoadCallback()
        loadCallback.callback()

        // Unloads the repository to prevent duplication.
        unload()

        loader.executor.execute {
            try {
                // Get the subject getSources from the getLoader which can come from a
                // a getClient or a getStore.
                val sources = loader.loadSubjects<Examination>(examinationClass)

                for (source in sources) {
                    var subject: Subject? = get(source.id)

                    if (subject != null) {
                        subject.sources.add(source)
                    } else {
                        subject = factory.create(source.id, source.name)
                        subject.sources.add(source)

                        put(subject)
                    }
                }
                _isLoaded = true

                // Call the callback on the callback getExecutor.
                loader.executor.callbackExecutor.execute {
                    loadCallback.onLoadCallback?.invoke()
                }
            } catch (e: Exception) {
                // Call the callback on the callback getExecutor.
                loader.executor.callbackExecutor.execute {
                    loadCallback.onErrorCallback?.invoke(e)
                }
            }
        }
    }

    /**
     * Unloads the [SubjectRepository].
     */
    override fun unload() {
        subjects.clear()
        _isLoaded = false
    }

    /**
     * Saves the [SubjectRepository].
     *
     * @throws Exception When an exception is thrown.
     */
    @Throws(Exception::class)
    override fun save() {
        // Merge all the subject getSources into a single linear list.
        val sources = ArrayList<SubjectSource>()
        for (subject in this) {
            for (source in subject.sources) {
                sources.add(source)
            }
        }

        // Pass the subjects to save to the getSaver instance.
        saver.saveSubjects<Examination>(examinationClass, sources)
    }

    /**
     * Saves the [SubjectRepository].
     *
     * @param callback Callback initializer.
     */
    override fun saveAsync(callback: Repository.SaveCallback.() -> Unit) {
        val saveCallback = Repository.SaveCallback()
        saveCallback.callback()

        val sources = ArrayList<SubjectSource>()
        for (subject in this) {
            for (source in subject.sources) {
                sources.add(source)
            }
        }

        saver.executor.execute {
            try {

                saver.saveSubjects<Examination>(examinationClass, sources)

                saver.executor.callbackExecutor.execute {
                    saveCallback.onSaveCallback?.invoke()
                }
            } catch (e: Exception) {
                saver.executor.callbackExecutor.execute {
                    saveCallback.onErrorCallback?.invoke(e)
                }
            }
        }
    }

    /**
     * Returns an iterator through the [Subject] in the [SubjectRepository].
     *
     * @return An iterator through the [Subject] in the [SubjectRepository].
     */
    override fun iterator(): MutableIterator<Subject> {
        return subjects.values.iterator()
    }
}
